### ISSUE TITLE
Inconsistent results with or without tf.function

### ISSUE THREAD

**Issue by njuaplusplus** [tensorflow/tensorflow#55814]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

tf 2.8.0

### Custom Code

No

### OS Platform and Distribution

Google Colab

### Current Behaviour?

```shell
The function with tf.function decorator should produce the same results the one without the decorator. In addition, adding one tf.print statement in the function with tf.function decorator will change the behavior.
```

### Standalone code to reproduce the issue

```shell
I created a google colab notebook. Please check this link.

https://colab.research.google.com/drive/1FOvjw0pPZHQw2bZtuFE8fnTAZk8_n0X8?usp=sharing
```

### Relevant log output

```shell
TensorFlow version: 2.8.0
loss_object tf.Tensor(26.064999, shape=(), dtype=float32)
loss_object_1 tf.Tensor(nan, shape=(), dtype=float32)
[[1.5240801e-14 4.78756791e-12 7.77465369e-15 ... 1 4.88135581e-29 0]]
loss_object_2 tf.Tensor(nan, shape=(), dtype=float32)
```
</details>

**Comment by tilakrayal** [comment]
@njuaplusplus ,
 tf.function does not provide bitwise identity of results. Floating point math is non-associative and non-commutable -- thus optimization passes can affect the result.
Different devices may actually yield different results for the same operation if they choose to implement FP operations differently.Thanks!

**Comment by njuaplusplus** [comment]
@tilakrayal 
Thank you for your reply. 
> tf.function does not provide bitwise identity of results. Floating point math is non-associative and non-commutable -- thus optimization passes can affect the result.

But does this indicate we should not use tf.function? Instead of calling this inconsistent results, I would rather say it's the *wrong* result. Do you mean this is not avoidable?

> Different devices may actually yield different results for the same operation if they choose to implement FP operations differently

Even running the code in the *same runtime* of one colab notebook gives different results. Does it mean even when we are connected to the same runtime, we are using different devices?

**Comment by tilakrayal** [comment]
@sachinprasadhs ,
I was able to reproduce the issue in tf v2.7, v2.8 and nightly.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/9d66143fceb5643b344a27a40f8f96aa/55814.ipynb).

**Comment by bhack** [comment]
I don't understand the `ŧf.print` subcase with `tf.float32` but it seems that is going to have some implicit precision impact on the Tensor itself /cc @mdanatg 

For the other behavior it seems that Softmax in graph mode has more precision than eager so with extreme values if could get `inf/-inf` or not in the two modes with `tf.float32`.

```python 
import tensorflow as tf

def softmax_eager():
    a = tf.nn.softmax(tf.constant([-1, 0.00001, 1.],dtype=tf.float32))
    print("Eager: " + str(a.numpy()))
    return a

@tf.function
def softmax_graph():
    a = tf.nn.softmax([-1, 0.00001, 1.])
    print("Graph:")
    tf.print(a)
    return a

print("Eager: " + str(softmax_eager().numpy()))
print("Grap: \n"  + str(softmax_graph().numpy()))
```

```python
Eager: [0.09003035 0.24473032 0.66523933]
Eager: [0.09003035 0.24473032 0.66523933]
Graph:
[0.0900303498 0.244730324 0.665239334]
Grap: 
[0.09003035 0.24473032 0.66523933]
```


If you test you example in `double/tf.float64` you will see that you have the same results in graph end eager mode (without or without `tf.print`):

```python
import numpy as np

import tensorflow as tf

print("TensorFlow version:", tf.__version__)

# download the input tensor
! wget "https://purdue0-my.sharepoint.com/:u:/g/personal/an93_purdue_edu/Eftt7AGnHMZNotkKjli-90ABboga38y0gmR9TyDZXyyn9w?e=aXRobr&download=1" -O "outputs.npy"

@tf.function
def loss_object(target, pred):
  target = tf.one_hot(target, 10, dtype=tf.float64)
  pred_conf = tf.nn.softmax(pred)
  cross_entropy = -tf.reduce_sum(target * tf.math.log(pred_conf))
  return cross_entropy

def loss_object_1(target, pred):
  target = tf.one_hot(target, 10, dtype=tf.float64)
  pred_conf = tf.nn.softmax(pred)
  cross_entropy = -tf.reduce_sum(target * tf.math.log(pred_conf))
  return cross_entropy

@tf.function
def loss_object_2(target, pred):
  target = tf.one_hot(target, 10, dtype=tf.float64)
  pred_conf = tf.nn.softmax(pred)
  tf.print(pred_conf)
  cross_entropy = -tf.reduce_sum(target * tf.math.log(pred_conf))
  return cross_entropy

outputs = np.load('./outputs.npy')
outputs = tf.convert_to_tensor(outputs, dtype=tf.float64)

print('loss_object', loss_object([1], outputs))
print('loss_object_1', loss_object_1([1], outputs))
print('loss_object_2', loss_object_2([1], outputs))
```

```python
loss_object tf.Tensor(26.064998269085926, shape=(), dtype=float64)
loss_object_1 tf.Tensor(26.064998269085926, shape=(), dtype=float64)
[[1.5240794106360253e-14 4.7875693825606735e-12 7.774643109731601e-15 ... 0.99999999999518918 4.8813465380882351e-29 2.7139525690757968e-40]]
loss_object_2 tf.Tensor(26.064998269085926, shape=(), dtype=float64)

```

**Comment by mdanatg** [comment]
That looks like an inconsistency between tf.print and print: tf.print shows 9 decimals, while Python shows just 8. Can you try with tf.print in eager mode?

**Comment by bhack** [comment]
Yes sorry I supposed to have a `tf.print` also in eager.. there was some copy past change. 
But can you see that with `tf.float64` we have the same results in eager and graph mode?

**Comment by mdanatg** [comment]
Not sure I see - you use prints in the second example, and the only tf.print outputs something else?
I think the numbers are equal in both cases (try verifying it) - the two prints only round the displayed value.

**Comment by bhack** [comment]
In the coalb in `tf.float32` we have these results:
```python
loss_object tf.Tensor(26.064999, shape=(), dtype=float32)
loss_object_1 tf.Tensor(nan, shape=(), dtype=float32)
[[1.5240801e-14 4.78756791e-12 7.77465369e-15 ... 1 4.88135581e-29 0]]
loss_object_2 tf.Tensor(nan, shape=(), dtype=float32)
```

In `tf.float64` we have:
```pyton
loss_object tf.Tensor(26.064998269085926, shape=(), dtype=float64)
loss_object_1 tf.Tensor(26.064998269085926, shape=(), dtype=float64)
[[1.5240794106360253e-14 4.7875693825606735e-12 7.774643109731601e-15 ... 0.99999999999518918 4.8813465380882351e-29 2.7139525690757968e-40]]
loss_object_2 tf.Tensor(26.064998269085926, shape=(), dtype=float64)
```

**Comment by njuaplusplus** [comment]
@bhack I see. Thank you very much. It's interesting that graph mode and eager mode have different precisions and the `tf.print` will affect the precision.  I previously thought the precision across all the computations will conform to the precision of the inputs.

**Comment by mdanatg** [comment]
I suspect the non-nan results for loss_object are thanks to arithmetic simplifications made by TF. Without those simplifications, you might hit numerical instability, especially in FP32. In eager, there are no such optimizations, and if you insert tf.print in graph mode, that would block such optimizations as well. Try disabling the optimizer to see if the inconsistency persists.

**Comment by bhack** [comment]
@mdanatg Yes, so it is still part of `Grappler` graph optimizzations. It is a little bit counterintuitive/under-exposed in docs but ok... We have all `nan` in `float32` now with:

```python
options =  tf.config.optimizer.get_experimental_options()
options['arithmetic_optimization']=False
tf.config.optimizer.set_experimental_options(options)`
```

```python
loss_object tf.Tensor(nan, shape=(), dtype=float32)
tf.Tensor([0.09003057 0.24472848 0.66524094], shape=(3,), dtype=float32)
loss_object_1 tf.Tensor(nan, shape=(), dtype=float32)
[0.0900305733 0.244728476 0.665240943]
loss_object_2 tf.Tensor(nan, shape=(), dtype=float32)
```

Can we close this? As this is enabled by default do we want to add a notice box in https://www.tensorflow.org/guide/function or something else?

**Comment by mdanatg** [comment]
I think it's safe to close, and we can think about where to best add a notice. Leaving the `tf.print` artifact aside (which is the kind of "careful not to confuse the compiler"caveat that we occasionally see in C++), I think in general this is another good reason to always develop for tf.function: add the tf.function early, debug and test with that, rather than starting in eager and adding tf.function later.

**Comment by bhack** [comment]
> I think in general this is another good reason to always develop for tf.function: add the tf.function early, debug and test with that, rather than starting in eager and adding tf.function later.

I agree but it is that generally in the Model this will happen automatically but in other cases you will need to do it manually. I don't know how this is really clear to many users/developers.
